dockerfile-image-update/src/main/java/com/salesforce/dockerfileimageupdate/utils/RateLimiter.java

```diff
@@ -0,0 +1,224 @@
+                throw new UnknownFormatConversionException("Can not tokenize/parse a null object");
+            }
+
+            long rLimit;
+            long duration;
+            Duration rLimitDuration;
+            Duration tokAddingRate;
+            String literalConstant;
+            String rateDuration;
+            String rateDurationUnitChar;
+
+            Matcher matcher = eventPattern.matcher(input);
+
+            if (!matcher.matches()) {
+                throw new UnknownFormatConversionException("input argument is not in valid format and can't be tokenize");
+            } else {
+                /**
+                 * eventPattern regex is divided into 5 groups
+                 * 0 group being the whole expression
+                 * 1 group is numeric rate limit value and will always
+                 *                          be present if exp matched
+                 * 2 is literal constant '-per-' and is optional
+                 * 3 is numeric rate duration and is optional
+                 * 4 is a char(s/m/h) and is optional
+                 */
+                rLimit = Long.parseLong(matcher.group(1)); // will always have a numeric value that fits in Long
+                literalConstant = matcher.group(2); // can be null or empty or '-per-'
+                rateDuration = matcher.group(3);// can be null or empty or numeric value
+                rateDurationUnitChar = matcher.group(4); //can be empty or char s,m,h
+            }
+
+            if (literalConstant == null || literalConstant.isEmpty()) {
+                //return rate with default duration and token adding rate
+                return new RateLimitEvent(rLimit, Constants.DEFAULT_RATE_LIMIT_DURATION, Constants.DEFAULT_TOKEN_ADDING_RATE);
+            }
+
+            // value is either going to be a valid number or null. defaulting to 1 unit(every hour/every min and so on)
+            duration = NumberUtils.isParsable(rateDuration) ? Long.parseLong(rateDuration) : 1;
+
+            switch (rateDurationUnitChar) {
+                case "s":
+                case "S":
+                    rLimitDuration = Duration.ofSeconds(duration);
+                    tokAddingRate = Duration.ofSeconds(rLimit / duration);
+                    break;
+                case "m":
+                case "M":
+                    rLimitDuration = Duration.ofMinutes(duration);
+                    tokAddingRate = Duration.ofMinutes(rLimit / duration);
+                    break;
+                case "h":
+                case "H":
+                    rLimitDuration = Duration.ofHours(duration);
+                    tokAddingRate = Duration.ofHours(rLimit / duration);
+                    break;
+                default:
+                    //should not reach here are regex will enforce char, keeping it for any unexpected use case.
+                    throw new UnknownFormatConversionException("Unexpected format encountered, can't tokenize input");
+            }
```

Comment:
Is there a reason why we are building our own duration parser instead of using Java's duration parsing APIs.  See https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/time/LocalTime.html#parse(java.lang.CharSequence) and https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/time/Duration.html#parse(java.lang.CharSequence) .  See https://stackoverflow.com/questions/8257641/java-how-to-convert-a-string-hhmmss-to-a-duration for details.

Reply:
we are asking customers to provide a non-ISO-8601 duration format that is more intuitive and contains multiple values in a string

Reply:
Why not just ask for the standard format?  We'd write less code and handle more cases.

Here is the translation of the test cases to use the ISO-8601 standard duration instead of our custom format.  I don't think the inputs look all that different.  But the parsing code shrinks to basically `Duration.parse(input.split("-")[1])`
```
500-per-s" -> 500-PT1S
"500-per-60s" -> 500-PT60S
"500-per-1m" -> 500-PT1M
"500-per-1h" -> 500-PT1H
"1234-per-2h - 1234-PT2H
```

Reply:
I like what @cjneasbi is recommending here - it is much cleaner and standardized

Reply:
ahh.. thought of that while implementing and decided to go with the exact format suggested in the previous code review comments. Let me make the change and notify you guys as soon as I get time.
thanks